### ISSUE TITLE
Allow aborting HTTP client connections in the _newclient API

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3811]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)| @radix reported|
|-|-|
|Trac ID|trac#3811|
|Type|enhancement|
|Created|2009-05-02 22:24:52Z|
|Branch|https://github.com/twisted/twisted/tree/expressive-client-timeouts-3811|

It should be possible to abort connections that are in progress in the _newclient API. This should close the connection and notify all application-callback APIs that their requests have been aborted.

<details><summary>Searchable metadata</summary>

```
trac-id__3811 3811
type__enhancement enhancement
reporter__radix radix
priority__normal normal
milestone__ 
branch__branches_expressive_client_timeouts_3811 branches/expressive-client-timeouts-3811
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__httpclient httpclient
time__1241303092000000 1241303092000000
changetime__1267069748000000 1267069748000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_1"></a>|@radix commented|
|-|-|

This is ready for review in the bzr branch at 

    http://twistedmatrix.com/users/radix/twisted-branches/expressive-client-timeouts/

This bzr branch is based on the branch expressive-http-client-886-3, which is accessible via bzr at:

    http://svn.twistedmatrix.com/bzr/Twisted/branches/expressive-http-client-886-3/


If another format would be preferable to a reviewer, please feel free to request it.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|ivank commented|
|-|-|

In twisted/web/test/test_newclient.py, the name `deliverBodyAndReturnResponse` is not defined. The tests don't pass.

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|ivank commented|
|-|-|

2. There should be vertical space after the end of `def deliverBody(response) `.

3. `from twisted.web._newclient import ChunkedEncoder, RequestGenerationFailed, RequestTransmissionFailed, ResponseFailed, WrongBodyLength, RequestNotSent, ConnectionAborted` is 169 characters. Perhaps it should be no longer than the second-longest import in test_newclient.py

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|ivank commented|
|-|-|

I can't review the necessity of the abort, so someone else should do that.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_6"></a>|@radix commented|
|-|-|

1.The NameError has been fixed.
 2. Fixed.
 3. I've resized this line to 80 columns. I didn't clean up *all* the >80 column lines in this branch to prevent unnecessary conflicts against the branch that this is based off of, since it hasn't yet been merged to trunk.

Thanks for the review.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph set owner to @radix|
|-|-|

It looks... okay, I guess?  I can't tell how to get a diff out of the branch.  Also, what exactly is the protocol for reviewing a branch based on another branch?  I guess if ... [#10610](https://github.com/twisted/twisted/issues/10610) lands without modification, then...

No, I don't know what to do.  This is outside the usual flow of things in too many different ways for me to deal with at once.  What I _think_ are the changes in this branch _appear_ be fine to be landed on trunk, if trunk were [#10610](https://github.com/twisted/twisted/issues/10610) already.  But I can't tell buildbot to do anything with it.

Unfortunately I think if I saw some serious problem with the code itself this review would be much easier.  Please make your expectations for the output of this review, and some directions for how the reviewer should proceed, a little clearer.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_8"></a>|@radix commented|
|-|-|

Okay, so here's how you'd review this change:

```
cd whatever
bzr branch http://svn.twistedmatrix.com/bzr/Twisted/branches/expressive-http-client-886-3/
cd expressive-http-client-886-3
bzr merge http://twistedmatrix.com/users/radix/twisted-branches/expressive-client-timeouts/
bzr diff | less
bzr revert
```

There are two things we could do with a +1: merge this branch into the expressive-http-client-886-3 timeout, or just leave it as a branch until expressive-http-client-886-3 is merged, and then merge this branch immediately after. Both of these options have some problems. The first one makes reviewing the final branch more confusing, or at least harder. The second one means maintaining divergences and resolving conflicts. I think we should do the second one, because hopefully we can get expressive-http-client merged to trunk quickly.

So I'm happy with a "+1, don't merge this branch".

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun removed owner|
|-|-|

[#10565](https://github.com/twisted/twisted/issues/10565) and [#3987](https://github.com/twisted/twisted/issues/3987) are in trunk now.  That should make the review process for this ticket pretty straightforward.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_10"></a>|@radix commented|
|-|-|

(In [27665]) Branching to 'expressive-client-timeouts-3811'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @radix|
|-|-|

This seems to be good, except it's missing a news topfile.  Add that, and if the [build](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/expressive-client-timeouts-3811) succeeds on all platforms then please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set status to `closed`|
|-|-|

(In [28528]) Merge expressive-client-timeouts-3811

Author: radix
Reviewer: ivank, glyph, exarkun
Fixes: [#3811](https://github.com/twisted/twisted/issues/3811)

Add `HTTP11ClientProtocol.abort` to interrupt a request/response early by dropping
the connection.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
